From 997687a90c1f57487b21e41b2d0ec39aec3f16e0 Mon Sep 17 00:00:00 2001
From: Anand N Sunkad <asunka@codeaurora.org>
Date: Mon, 20 Oct 2014 18:37:27 +0530
Subject: [PATCH 26/31] wcnss: configure IRIS clock to 19.2Mhz

Set IRIS clock  to 19.2Mhz if there is no iris
card found or detected IRIS card's chip id is invalid.

Change-Id: Ib9063e67315faca0a3817c1c5df8655f57b9c381
CRs-fixed: 743022
Signed-off-by: Anand N Sunkad <asunka@codeaurora.org>
---
 drivers/net/wireless/wcnss/wcnss_vreg.c | 30 ++++++++++++++++++++++++++++--
 1 file changed, 28 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/wcnss/wcnss_vreg.c b/drivers/net/wireless/wcnss/wcnss_vreg.c
index 6cfd8cb..6af1a9a 100644
--- a/drivers/net/wireless/wcnss/wcnss_vreg.c
+++ b/drivers/net/wireless/wcnss/wcnss_vreg.c
@@ -44,6 +44,13 @@ static int auto_detect;
 
 #define PRONTO_IRIS_REG_READ_OFFSET       0x1134
 #define PRONTO_IRIS_REG_CHIP_ID           0x04
+/* IRIS card chip ID's */
+#define WCN3660       0x0200
+#define WCN3660A      0x0300
+#define WCN3660B      0x0400
+#define WCN3620       0x5111
+#define WCN3620A      0x5112
+#define WCN3610       0x9101
 
 #define WCNSS_PMU_CFG_IRIS_XO_CFG          BIT(3)
 #define WCNSS_PMU_CFG_IRIS_XO_EN           BIT(4)
@@ -141,6 +148,24 @@ int xo_auto_detect(u32 reg)
 	}
 }
 
+int validate_iris_chip_id(u32 reg)
+{
+	int iris_id;
+	iris_id = reg >> 16;
+
+	switch (iris_id) {
+	case WCN3660:
+	case WCN3660A:
+	case WCN3660B:
+	case WCN3620:
+	case WCN3620A:
+	case WCN3610:
+		return 0;
+	default:
+		return 1;
+	}
+}
+
 static int configure_iris_xo(struct device *dev, bool use_48mhz_xo, int on,
 			int *iris_xo_set)
 {
@@ -235,8 +260,9 @@ static int configure_iris_xo(struct device *dev, bool use_48mhz_xo, int on,
 
 			iris_reg = readl_relaxed(iris_read_reg);
 			pr_info("wcnss: IRIS Reg: %08x\n", iris_reg);
-			if (iris_reg == PRONTO_IRIS_REG_CHIP_ID) {
-				pr_info("wcnss: IRIS Card not Preset\n");
+
+			if (validate_iris_chip_id(iris_reg)) {
+				pr_info("wcnss: IRIS Card absent/invalid\n");
 				auto_detect = WCNSS_XO_INVALID;
 				/* Reset iris read bit */
 				reg &= ~WCNSS_PMU_CFG_IRIS_XO_READ;
-- 
2.5.0

